kod/object/active/holder/room/monsroom/survivalroom.kod

```diff
@@ -0,0 +1,1299 @@
+            }
+            
+            if plGenerators <> $
+            {
+               lRandomSpawnPoint = Nth(plGenerators,
+                                       Random(1,Length(plGenerators)));
+            }
+            else
+            {
+               lRandomSpawnPoint = [Send(self,@GetTeleportRow),
+                                    Send(self,@GetTeleportCol)];
+            }
+            
+            Send(self,@NewHold,#what=oReward,
+                               #new_row=Nth(lRandomSpawnPoint,1),
+                               #new_col=Nth(lRandomSpawnPoint,2));
+         }
+      }
+      
+      piLevel = piLevel + 1;
+      
+      return;
+   }
+   
+   ChooseGoals()
+   {
+      local iRand,plBosses,iNumAdmins,i;
+      
+      iNumAdmins = 0;
+      for i in plParticipants
+      {
+         if IsClass(i,&DM)
+         {
+            iNumAdmins = iNumAdmins+1;
+         }
+      }
+      
+      iRand = Random((Length(plParticipants)-iNumAdmins)*10,
+                     (Length(plParticipants)-iNumAdmins)*30);
+                     
+      iRand = Bound(iRand,1,150);
+
+      plLevelGoals = Cons([GOAL_KILLS,iRand],plLevelGoals);
+
+      iRand = Random(1,3);
+      if iRand = 1
+         AND piSpawnedBossThisRound
+      {
+         plLevelGoals = Cons([GOAL_BOSS,1],plLevelGoals);
+      }
+
+      if iRand = 2
+      {
+         plLevelGoals = Cons([GOAL_MINIBOSSES,2],plLevelGoals);
+      }
+
+      iRand = Random(1,5);
+      if iRand = 1
+         AND piSpawnedBossThisRound
+      {
+         plExtraLifeGoals = Cons([GOAL_BOSS,1],plExtraLifeGoals);
+      }
+      if iRand = 2
+      {
+         plExtraLifeGoals = Cons([GOAL_MINIBOSSES,2],plExtraLifeGoals);
+      }
+      if iRand = 3
+         AND (piPublic
+         OR poGuildAssociation = $)
+      {
+         plExtraLifeGoals = Cons([GOAL_KILL_FRIENDLY,1],plExtraLifeGoals);
+      }
+      
+      return;
+   }
+   
+   SpawnBoss()
+   {
+      local cBoss;
+
+      cBoss = Nth(plBosses,Random(1,Length(plBosses)));
+      Send(self,@GenerateMonster,#oMonster=Create(cBoss),#bStack=TRUE,
+                                    #piSurvivalLevel=piLevel);
+      piSpawnedBossThisRound = TRUE;
+
+      return;
+   }
+   
+   RespawnMiniboss(timer=$)
+   {
+      ptRespawnMiniBossTimer = $;
+      Send(self,@SpawnMiniboss);
+
+      return;
+   }
+   
+   SpawnMiniboss()
+   {
+      local cMiniboss;
+
+      cMiniboss = Nth(plMinibosses,Random(1,Length(plMinibosses)));
+      Send(self,@GenerateMonster,#oMonster=Create(cMiniboss),#bStack=TRUE,
+                                 #piSurvivalLevel=piLevel);
+      return;
+   }
+   
+   SomethingKilled(what=$,victim=$)
+   {
+      local i, n, p, cDeadClass, respawnCheck;
+      
+      if IsClass(victim,&User)
+      {
+         propagate;
+      }
+      
+      % Minions!
+      if (IsClass(victim,&Monster)
+         AND Send(what,@GetMaster) <> $)
```

Comment:
Send(what,@GetMaster) <> $ would catch cases where a minion was incorrectly given credit for a kill, however KilledSomething in Monster should be catching every case of this (the monster evil twins were an exception because they don't have a master) and every battler that sends SomethingKilled to the victim's owner (i.e. this room) should have a $ master (because the message will recurse until it finds a battler with no master).

To catch minions killed use Send(victim,@GetMaster) <> $ which should cover reflections, ETs, seduced and animated mobs. You should also add a check for Send(victim,@IsIllusion) to catch apparitions and monster ETs.